src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,424 @@
+else:
+    from typing_extensions import Literal
+
+import requests
+
+from .constants import ENDPOINT
+from .lfs import UploadInfo, lfs_upload, post_lfs_batch_info, validate_batch_actions
+from .utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+UploadMode = Literal["lfs", "regular"]
+
+
+@dataclass
+class CommitOperationDelete:
+    """
+    Data structure holding necessary info to delete
+    a file from a repository on the Hub
+
+    Args:
+        path_in_repo (`str`):
+            Relative filepath in the repo, for example:
+            `"checkpoints/1fec34a/weights.bin"`
+    """
+
+    path_in_repo: str
+
+
+@dataclass
+class CommitOperationAdd:
+    """
+    Data structure holding necessary info to upload a file
+    to a repository on the Hub
+
+    Args:
+        path_in_repo (`str`):
+            Relative filepath in the repo, for example:
+            `"checkpoints/1fec34a/weights.bin"`
+        path_or_fileobj (`str`, `bytes`, or `BinaryIO`):
+            Either:
+            - a path to a local file (as str) to upload
+            - a buffer of bytes (`bytes`) holding the content of the file to upload
+            - a "file object" (subclass of `io.BufferedIOBase`), typically obtained
+                with `open(path, "rb")`
+    """
+
+    path_in_repo: str
+    path_or_fileobj: Union[str, bytes, BinaryIO]
+
+    _upload_info: Optional[UploadInfo] = field(default=None, init=False)
+
+    def validate(self):
+        """
+        Ensures `path_or_fileobj` is valid.
+
+        Raises: `ValueError`
+        """
+        if isinstance(self.path_or_fileobj, str):
+            path_or_fileobj = os.path.normpath(os.path.expanduser(self.path_or_fileobj))
+            if not os.path.isfile(path_or_fileobj):
+                raise ValueError(
+                    f"Provided path: '{path_or_fileobj}' is not a file on the local"
+                    " file system"
+                )
+        elif not isinstance(self.path_or_fileobj, (io.BufferedIOBase, bytes)):
+            # ^^ Inspired from: https://stackoverflow.com/questions/44584829/how-to-determine-if-file-is-opened-in-binary-or-text-mode
+            raise ValueError(
+                "path_or_fileobj must be either an instance of str, bytes or"
+                " io.BufferedIOBase. If you passed a file-like object, make sure it is"
+                " in binary mode."
+            )
+        if isinstance(self.path_or_fileobj, io.BufferedIOBase):
+            try:
+                self.path_or_fileobj.tell()
+                self.path_or_fileobj.seek(0, os.SEEK_CUR)
+            except (OSError, AttributeError) as exc:
+                raise ValueError(
+                    "path_or_fileobj is a file-like object but does not implement"
+                    " seek() and tell()"
+                ) from exc
+
+    def upload_info(self) -> UploadInfo:
+        """
+        Computes and caches UploadInfo for the underlying data behind `path_or_fileobj`
```

Comment:
would be nice to have hyperlinks to objects when we refer to them in the docs (`UploadInfo`)

Also, I'm not sure if I understand what this method does if I was reading the docstring out of context. Is this a method which is supposed to be called by users? i.e. should it be public or private?

Reply:
Made it private